nucypher/blockchain/eth/sol/source/contracts/Issuer.sol

```diff
@@ -56,11 +56,20 @@ contract Issuer is Upgradeable {
     )
         public
     {
-        require(_token.totalSupply() > 0 &&
+        totalSupply = _token.totalSupply();
+        require(totalSupply > 0 &&
             _miningCoefficient != 0 &&
             _hoursPerPeriod != 0 &&
             _lockedPeriodsCoefficient != 0 &&
             _rewardedPeriods != 0);
+        uint256 maxLockedPeriods = _rewardedPeriods + _lockedPeriodsCoefficient;
+        require(maxLockedPeriods > _rewardedPeriods &&
+            // worst case for `totalLockedValue * k2`, when totalLockedValue == totalSupply
+            totalSupply * miningCoefficient / totalSupply == miningCoefficient &&
+            // worst case for `(totalSupply - currentSupply) * lockedValue * (k1 + allLockedPeriods)`,
+            // when currentSupply == 0, lockedValue == totalSupply
+            totalSupply * totalSupply * maxLockedPeriods / totalSupply / totalSupply == maxLockedPeriods,
```

Comment:
I don't understand what these checks are doing... won't `totalSupply`always cancel out to `maxLockedPeriods == maxLockedPeriods` and `miningCoefficient == miningCoefficient`? So when can this require fail?

Reply:
It will fail in case of overflow after multiplication, `totalSupply * totalSupply * maxLockedPeriods` will be less than `totalSupply` or `maxLockedPeriods`